### ISSUE TITLE
[DataGridPro] styleOverrides does not support Theme typings

### ISSUE THREAD

**Issue by iigrik** [mui/mui-x#5816]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Usage of `theme` object in callback of styleOverrides causing ts error: Object is of type 'unknown'.

```ts
const theme = createTheme({
  components: {
    MuiDataGrid: {
      styleOverrides: {
        root: ({ theme }) => ({
          backgroundColor: theme.palette.primary.main
        })
      }
    }
  }
});
```

### Expected behavior 🤔

There are no ts errors presented. `theme` object has correct typing.

### Steps to reproduce 🕹

Link to live example: https://codesandbox.io/s/datagridpro-styleoverrides-issue-0k8c4r

**Comment by minyan-gu** [comment]
noticed same issue, would be nice to have a patch release soon.

**Comment by oliviertassinari** [comment]
Makes sense, this has changed after https://github.com/mui/material-ui/pull/30524, e.g.

https://github.com/siriwatknp/material-ui/blob/f981c6c2dfd4babfd881fc69912fd8c513d33f2c/packages/mui-lab/src/themeAugmentation/components.d.ts#L16

**Comment by iigrik** [comment]
@oliviertassinari Should I change/add something to the linked PRs? Is implementation correct?

**Comment by iigrik** [comment]
@oliviertassinari This issue also presented in "DataPickerPro" component. Can I create PR in scope of current issue or separate issue should be created?
